SimCalorimetry/HGCalAssociatorProducers/plugins/TracksterAssociatorByEnergyScoreImpl.cc

```diff
@@ -0,0 +1,519 @@
+    double tot_energy = 0.;
+    for (auto const& haf : tssInSimCluster[sc].hits_and_fractions) {
+      LogDebug("TracksterAssociatorByEnergyScoreImpl")
+          << "      Hits/fraction/energy: " << (uint32_t)haf.first << "/" << haf.second << "/"
+          << haf.second * hitMap_->at(haf.first)->energy() << std::endl;
+      tot_energy += haf.second * hitMap_->at(haf.first)->energy();
+    }
+    LogDebug("TracksterAssociatorByEnergyScoreImpl") << "    Tot Sum haf: " << tot_energy << std::endl;
+    for (auto const& ts : tssInSimCluster[sc].tracksterIdToEnergyAndScore) {
+      LogDebug("TracksterAssociatorByEnergyScoreImpl")
+          << "      tsIdx/energy/score: " << ts.first << "/" << ts.second.first << "/" << ts.second.second << std::endl;
+    }
+  }
+
+  LogDebug("TracksterAssociatorByEnergyScoreImpl") << "Improved detIdToSimClusterId_Map INFO" << std::endl;
+  for (auto const& sc : detIdToSimClusterId_Map) {
+    LogDebug("TracksterAssociatorByEnergyScoreImpl")
+        << "For detId: " << (uint32_t)sc.first
+        << " we have found the following connections with SimClusters:" << std::endl;
+    for (auto const& sclu : sc.second) {
+      LogDebug("TracksterAssociatorByEnergyScoreImpl")
+          << "  SimCluster Id: " << sclu.clusterId << " with fraction: " << sclu.fraction
+          << " and energy: " << sclu.fraction * hitMap_->at(sc.first)->energy() << std::endl;
+    }
+  }
+#endif
+
+  // Update scsInTrackster; compute the score Trackster-to-SimCluster,
+  // together with the returned AssociationMap
+  for (unsigned int tsId = 0; tsId < nTracksters; ++tsId) {
+    // The SimClusters contributing to the Trackster's LayerClusters should already be unique.
+    // find the unique SimClusters id contributing to the Trackster's LayerClusters
+    std::sort(scsInTrackster[tsId].begin(), scsInTrackster[tsId].end());
+    auto last = std::unique(scsInTrackster[tsId].begin(), scsInTrackster[tsId].end());
+    scsInTrackster[tsId].erase(last, scsInTrackster[tsId].end());
+
+    // If a reconstructed Trackster has energy 0 but is linked to a
+    // SimCluster, assigned score 1
+    if (tracksters[tsId].raw_energy() == 0. && !scsInTrackster[tsId].empty()) {
+      for (auto& scPair : scsInTrackster[tsId]) {
+        scPair.second = 1.;
+        LogDebug("TracksterAssociatorByEnergyScoreImpl") << "TracksterId : \t " << tsId << "\t SC id : \t"
+                                                         << scPair.first << "\t score \t " << scPair.second << "\n";
+      }
+      continue;
+    }
+
+    float invTracksterEnergyWeight = 0.f;
+    for (const auto& lcId : tracksters[tsId].vertices()) {
+      const auto& hits_and_fractions = layerClusters[lcId].hitsAndFractions();
+
+      // Compute the correct normalization
+      for (auto const& haf : hits_and_fractions) {
+        invTracksterEnergyWeight +=
+            (haf.second * hitMap_->at(haf.first)->energy()) * (haf.second * hitMap_->at(haf.first)->energy());
```

Comment:
Aren't we missing the *multiplicity* usage of the `layerCluster` among different `tracksters`?
Here and, possibly, in other places in this `associator`?